### PR TITLE
Move destruct to own module

### PR DESCRIPTION
**Pull request by rgrinberg**
We choose the name destruct_lsp because toplevel destruct is taken by merlin

### CODE REVIEW

**Review by ulugbekna — COMMENTED** [review]
Thanks for the PR! It overall looks good to me. A couple of points:
- `_lsp` suffix isn't very helpful I feel in `destruct_lsp` and less in accord with other request handler modules like `compl`
- I would move all handling of this particular request to its own module as it's done with `switch_impl_intf`, so that we keep `ocaml_lsp_server.ml` cleaner

I would in general want to come up with maybe an interface for request handlers like 

```ocaml
module type RequestHandler = sig 
  type request 
  type response
  val on_request : state -> ~params:request -> response 
end
```

but I can do that in another PR.

**Comment by rgrinberg** [comment]
> _lsp suffix isn't very helpful I feel in destruct_lsp and less in accord with other request handler modules like compl

It's just a temporary workaround until I finish https://github.com/ocaml/dune/pull/3825 in dune. It's only because merlin is taking the `Destruct` name globally.

> I would move all handling of this particular request to its own module as it's done with switch_impl_intf, so that we keep ocaml_lsp_server.ml cleaner

Yes, but it's not clear to me how much general code action handling code should live in the main module. We'll get a better idea how to generalize once we have a second action.

> I would in general want to come up with maybe an interface for request handlers like

Isn't just a function enough? On the other hand, something for code actions could be useful.

File: ocaml-lsp-server/src/destruct_lsp.ml [hunk 1]
```diff
@@ -0,0 +1,34 @@
+open Import
+
+let action = "destruct"
```
[the remark this answers is not in the document]
Comment (rgrinberg) [reply ocaml-lsp-server/src/destruct_lsp.ml]: I chose `action_kind`. `kind` is a bit vague

**Comment by ulugbekna** [comment]
> Yes, but it's not clear to me how much general code action handling code should live in the main module. We'll get a better idea how to generalize once we have a second action.

Fair enough, but I feel like trivialities like param parsing belong to request handling logic and hence to the module. I would expect the request handlers to accept `state`, the request params, and return a response. 

```ocaml
    let open Fiber.Result.O in
    let uri = Uri.t_of_yojson (`String params.textDocument.uri) in
    let* doc = Fiber.return (Document_store.get store uri) in
    let+ action = Destruct_lsp.code_action doc params in
    let action =
      Option.map action ~f:(fun destruct -> [ `CodeAction destruct ])
    in
    (action, state)
```

Anyway, you're way more experienced than I am, so abstractions are up to your discretion :-)

> Isn't just a function enough? On the other hand, something for code actions could be useful.

I had in mind that we can have a general request handler interface for modules that specifically handle certain requests. We can then specialize/extend that interface for specific kinds of requests, I guess.

**Comment by rgrinberg** [comment]
Yes, that's a good point. On the other hand, I want to impose some restrictions on common request handlers and code actions:

* Modification of the server state isn't allowed. Or at the very least, it should be obvious from the type.
* Sanity checks like making sure that the document exists should be done in one place. These are duplicated all over the place and it's a little ugly.

I'm not entirely sure what's the correct way to organize things. For now, let's keep writing more actions & request handlers and see where that leads us.

**Pull request merged**